rst_files/lake_model.rst

```diff
@@ -669,84 +641,114 @@ Following :cite:`davis2006flow`, we set :math:`\alpha`, the hazard rate of leavi
 Fiscal Policy Code
 -----------------------
 
-We will make use of code we wrote in the :doc:`McCall model lecture <mccall_model>`, embedded below for convenience
+We will make use of code we wrote in the :doc:`McCall model lecture <mccall_model>`
+
+.. code-block:: julia
+
+    using Distributions, LinearAlgebra, Compat, Expectations, Parameters, NLsolve, Plots
+
+    function solve_mccall_model(mcm; U_iv = 1.0, V_iv = ones(length(mcm.w)), tol = 1e-5, 
+                                iter = 2_000)
+        # α, β, σ, c, γ, w = mcm.α, mcm.β, mcm.σ, mcm.c, mcm.γ, mcm.w
+        @unpack α, β, σ, c, γ, w, dist, u = mcm 
+        
+        # parameter validation
+        @assert c > 0.0
+        @assert minimum(w) > 0.0 # perhaps not strictly necessary, but useful here 
+        
+        # necessary objects 
+        u_w = u.(w, σ) 
+        u_c = u(c, σ)
+        E = expectation(dist) # expectation operator for wage distribution
+
+        # Bellman operator T. Fixed point is x* s.t. T(x*) = x*
+        function T(x)
+            V = x[1:end-1]
+            U = x[end]
+            [u_w + β * ((1 - α) * V .+ α * U); u_c + β * (1 - γ) * U + β * γ * E * max.(U, V)]
+        end 
+
+        # value function iteration  
+        x_iv = [V_iv; U_iv] # initial x val
+        xstar = fixedpoint(T, x_iv, iterations = iter, xtol = tol).zero 
+        V = xstar[1:end-1]
+        U = xstar[end]
+        
+        # compute the reservation wage 
+        wbarindex = searchsortedfirst(V .- U, 0.0)
+        if wbarindex >= length(w) # if this is true, you never want to accept
```

Comment:
Could be an `ifelse`

Reply:
Is there a benefit in readability or consistency with using `ifelse` there?

Reply:
For a simple `ifelse` is easier to read
```
x = ifelse(cond, val_true, val_false)
```
vs
```
if cond
    x = val_true
else
    x = val_false
end
```

Reply:
So in this particular example
```julia
    if wbarindex >= length(w)
        w_bar = Inf
    else
        w_bar = w[wbarindex] # otherwise, return the number
    end
```
becomes

```
 # if index doesn't exist, then w_bar = Inf since they never accept
w_bar = ifelse(wbarindex >= length(w), Inf, w[barindex])
```
I am not sure what is more readable for intro users.  Close call.

Reply:
Personally, I would use
```
w_bar = wbarindex ≥ length(w) ? Inf : w[wbarindex]
```
since we lifted the ban on the ternary operator and seems pretty easy to follow.

Reply:
When did we lift the ternary ban?  The current guideline is identical to before
> Minimize use of the ternary operator. It is confusing for new users, so use it judiciously, and never purely to make code more terse.

This is a case where the ternary operator may be relatively readable since it follows a simple pattern... but we should always be suspicious.  It is worst when used nested in other functions.  But if we move to this, we need a comment explaining the ternary operator (each time we use it, since people will read these thigns out of orderr).

All that said, the `if` approach is so clear that I don't see the benefit of changing things like this.  Our time is spent worrying about first-order readability changes instead of wondering if we can make it a little more concise at the cost of using an operator we need to explain each time we use it.

Reply:
I thought we had allowed it only for simple cases. Guess I was wrong. It is pretty common for languages economists might use. MATLAB, R, Python, Stata, and SAS have it. Nvm.

Reply:
I think this is fine as-is.